File: smells/unrealistic/fantasy.js
/* Fantasy Tests
 *
 * Passing tests of code that wouldn't actually work in production
 */

// Тестируемый модуль
var authorizor = {
  roles: function (user, password) {
    if (authenticator.login({user: user, password: password})) {
      return ['admin', 'developer', 'manager']
    } else {
      return []
    }
  }
}

// Тесты
var td = require('testdouble')
module.exports = {
  beforeEach: function () {
    td.replace(authenticator, 'login')
  },
  allRolesIfAuthenticated: function () {
    td.when(authenticator.login({user: 'hi', password: 'bye'})).thenReturn(true)

    var result = authorizor.roles('hi', 'bye')

    assert.deepEqual(result, ['admin', 'developer', 'manager'])
  },
  noRolesIfNotAuthenticated: function () {
    td.when(authenticator.login({user: 'hi', password: 'no!'})).thenReturn(false)

    var result = authorizor.roles('hi', 'no!')

    assert.deepEqual(result, [])
  },
  afterEach: function () {
    td.reset()
  }
}

// Фейковая реализация
var authenticator = {
  login: function (credentials) {
    if (!credentials.password || !credentials['2fa']) {
      throw new Error('Both password and two-factor auth token are now required!')
    }
    return true
  }
}

File: smells/unrealistic/fantasy/fantasy.rb
# Subject under test
class Authorizer
  def initialize(authenticator)
    @authenticator = authenticator
  end

  def roles(user, password)
    if @authenticator.login(user: user, password: password)
      ["admin", "developer", "manager"]
    else
      []
    end
  end
end

# Test
class Fantasy < SmellTest
  def setup
    @authenticator = Minitest::Mock.new(Authenticator.new)
    @subject = Authorizer.new(@authenticator)
    super
  end

  def test_all_roles_if_authenticated
    @authenticator.expect(:login, true, [{user: "hi", password: "bye"}])

    result = @subject.roles("hi", "bye")

    assert_equal ["admin", "developer", "manager"], result
  end

  def test_no_roles_if_not_authenticated
    @authenticator.expect(:login, false, [{user: "hi", password: "no!"}])

    result = @subject.roles("hi", "no!")

    assert_equal [], result
  end
end

# Fake production implementations to simplify example test of subject
class Authenticator
  def login(credentials)
    if !credentials[:password] || !credentials[:"2fa"]
      raise "Both password and two-factor auth token are now required!"
    end
    return true
  end
end
